### PR TITLE
Dev 245

### PR DESCRIPTION
**Pull request by Palabola**
## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling with specific messages for different HTTP status codes.
	- Introduced a new toast notification component to display error messages visually.
	- Improved HTML rendering of server property data, allowing for more flexible presentation.
	- Added new input and output properties to the FAQ component for improved interactivity.

- **Style**
	- Added new CSS styles for the toast notification to improve visibility and transition effects.

### CODE REVIEW

File: src/app/pages/server-details/server-details.component.ts [hunk 1]
```diff
@@ -762,6 +773,24 @@ export class ServerDetailsComponent implements OnInit, OnDestroy {
       return undefined;
     }
 
+    if(name === 'storages') {
+      let html = '<ul>';
+      (prop as any[]).forEach((s: any, index: number) => {
+        html += `<li>${s.size}GB ${s.storage_type ? s.storage_type : ''}</li>`;
```
Comment (daroczig) [inline src/app/pages/server-details/server-details.component.ts:779]: there should be a space between the value and unit. same for `MB` below
```suggestion
        html += `<li>${s.size} GB ${s.storage_type ? s.storage_type : ''}</li>`;
```

File: src/app/pages/server-details/server-details.component.scss [hunk 2]
```diff
@@ -58,6 +58,22 @@
...
+  transition: transform 0.5s ease-out;
+  display: none;
+  /* other styles... */
+}
+
+#toast-danger.show {
+  transform: translateX(0);
+  display: flex;
+}
+
```
Comment (daroczig) [inline src/app/pages/server-details/server-details.component.scss:76]: IMO we need to add CSS for the `li` tags as well to make it visible that it's a list:

image

File: src/app/pages/server-details/server-details.component.ts [hunk 3]
```diff
@@ -762,6 +773,24 @@ export class ServerDetailsComponent implements OnInit, OnDestroy {
...
+      });
+      html += '</ul>';
+      return html;
+    }
+
+    if(name === 'gpus') {
+      let html = '<ul>';
+      (prop as any[]).forEach((s: any, index: number) => {
+        html += `<li>${s.model}${s.manufacturer ? (' (' + s.manufacturer + ')') : ''}${s.family ? ' family: ' + s.family + ',' : ''}${s.memory ? ' memory: ' + s.memory + 'MB' : ''}</li>`;
```
Comment (daroczig) [inline src/app/pages/server-details/server-details.component.ts:788]: If we are formatting this nicely, then I'd construct this as manufacturer (if available), then family (if available), then the exact model (if available) -- see how it's done at the FAQ specs already:

image

Then show the following fields in parentheses, but note the need for the space between the value and unit (like in all cases everywhere in the app):

1. memory amount
2. `Firmware version` if know
3. `BIOS version` if known
4. `Clock rate` (filled in by `graphics_clock`) if available

File: src/app/pages/server-details/server-details.component.html [hunk 4]
```diff
@@ -541,3 +541,21 @@ <h3 class="text-2xl" >Similar Servers </h3>
   class="absolute z-10 inline-block px-3 py-2 text-sm font-medium text-white transition-opacity duration-300 bg-secondary border border-solid border-emerald-400 rounded-lg">
   <div [innerHTML]="geekbenchHTML" ></div>
 </div>
+
+<div
+  id="toast-danger" #toastDanger
+  class="flex items-center w-full max-w-xs p-4 mb-4 text-white bg-secondary rounded-lg border boder-solid border-emerald-400" role="alert">
```
Comment (daroczig) [inline src/app/pages/server-details/server-details.component.html:547]: IMO we should use the warning/error colors provided by Szucsi: maybe try the orange as the background and the red as border?

File: src/app/pages/server-details/server-details.component.html [hunk 5]
```diff
@@ -541,3 +541,21 @@ <h3 class="text-2xl" >Similar Servers </h3>
...
+  <div class="inline-flex items-center justify-center flex-shrink-0 w-8 h-8 text-red-500 bg-red-100 rounded-lg dark:bg-red-800 dark:text-red-200">
+      <svg class="w-5 h-5" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 20 20">
+          <path d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5Zm3.707 11.793a1 1 0 1 1-1.414 1.414L10 11.414l-2.293 2.293a1 1 0 0 1-1.414-1.414L8.586 10 6.293 7.707a1 1 0 0 1 1.414-1.414L10 8.586l2.293-2.293a1 1 0 0 1 1.414 1.414L11.414 10l2.293 2.293Z"/>
```
Comment (daroczig) [inline src/app/pages/server-details/server-details.component.html:550]: we have a lucide icon for this, why not use that? `circle-x`

**Comment by daroczig** [comment]
why does the ul/li start left to the other items?

image

I'd expect the bullet points to be synced up with the words.

**Comment by daroczig** [comment]
When firmware version not available, we show and empty string between commas:

image

https://deploy-preview-67--creative-choux-a3c817.netlify.app/server/aws/p3.16xlarge?showDetails=true

**Comment by daroczig** [comment]
I don't see the `circle-x` icon being used:

image

https://lucide.dev/icons/circle-x

**Comment by daroczig** [comment]
DEV-207 is not only about the server details widget .. any chance we can support the others as well, e.g. FAQs?

File: src/app/pages/server-details/server-details.component.ts [hunk 6]
```diff
@@ -140,11 +144,14 @@ export class ServerDetailsComponent implements OnInit, OnDestroy {
               private SEOHandler: SeoHandlerService,
               private serverCompare: ServerCompareService,
               private router: Router,
+              private renderer: Renderer2,
```
[the remark this answers is not in the document]
Comment (Palabola) [reply src/app/pages/server-details/server-details.component.ts:147]: done

File: src/app/pages/server-details/server-details.component.ts [hunk 7]
```diff
@@ -1232,6 +1280,9 @@ export class ServerDetailsComponent implements OnInit, OnDestroy {
     }
 
     this.SEOHandler.setupStructuredData(this.document, [JSON.stringify(json)]);
+  }
 
+  showToast() {
+    this.renderer.addClass(this.toastDanger.nativeElement, 'show');
```
[the remark this answers is not in the document]
Comment (Palabola) [reply src/app/pages/server-details/server-details.component.ts:1286]: done

File: src/app/pages/server-details/server-details.component.ts [hunk 8]
```diff
@@ -442,7 +456,13 @@ export class ServerDetailsComponent implements OnInit, OnDestroy {
...
       }).catch((error) => {
         console.error('Failed to load server data:', error);
-        this.router.navigateByUrl('/');
+        if(error?.status === 404) {
+          this.toastErrorMsg = 'Server not found. Please try again later.';
+        }
+        if(error?.status === 500) {
+          this.toastErrorMsg = 'Internal server error. Please try again later.';
+        }
+        this.showToast();
```
[the remark this answers is not in the document]
Comment (Palabola) [reply src/app/pages/server-details/server-details.component.ts:465]: done

**Comment by Palabola** [comment]
> I don't see the `circle-x` icon being used:
> 
> image
> 
> https://lucide.dev/icons/circle-x

It looks better this way IMO.

**Comment by daroczig** [comment]
sorry, the comma issue is still happening with other combinations, see eg

image


https://deploy-preview-67--creative-choux-a3c817.netlify.app/server/aws/g4ad.16xlarge?openFAQ=2

maybe try creating an array of the available labels and join by `, `?

File: src/app/pages/server-details/server-details.component.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (daroczig) [reply src/app/pages/server-details/server-details.component.ts:1311]: actually, I think this is casing more harm than good: currently, SSR also waits 10 seconds for the popup to disappear, so page load is super slow (10 seconds!)

so at least keep this to the client-side, but IMO we can drop the hide altogether -- it's a permanent error on the page ... maybe hide when navigating away

[the remark this answers is not in the document]
Comment (Palabola) [reply src/app/pages/server-details/server-details.component.ts:1311]: Great idea, removed toastr from server-side. 
The toaster destroys when navigating from the page. 
For some reason the 404 SSR pageload is super slow even without the toastr. I suggest keep this in mind and not optimize for 404 load speed for now.

**Review by daroczig — APPROVED** [review]

**Pull request merged**